### Project 1: BRFSS Mental Health Analysis
#### Notebook 04: Regression Modeling
**Author:** Amanda Shero  
**Date:** April 2026  
**Description:** Regression modeling.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [2]:
df = pd.read_csv('df_encoded.csv')

In [3]:
df.shape

(425215, 45)

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 425215 entries, 0 to 425214
Data columns (total 45 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   _AGEG5YR         425215 non-null  float64
 1   SEXVAR           425215 non-null  float64
 2   EDUCA            425215 non-null  float64
 3   INCOME3          425215 non-null  float64
 4   _URBSTAT         425215 non-null  float64
 5   GENHLTH          425215 non-null  float64
 6   PHYSHLTH         425215 non-null  float64
 7   ADDEPEV3         425215 non-null  float64
 8   CVDINFR4         425215 non-null  float64
 9   ASTHMA3          425215 non-null  float64
 10  HAVARTH4         425215 non-null  float64
 11  CHCCOPD3         425215 non-null  float64
 12  EXERANY2         425215 non-null  float64
 13  SMOKDAY2         425215 non-null  float64
 14  AVEDRNK3         425215 non-null  float64
 15  DRNK3GE5         425215 non-null  float64
 16  _HLTHPL1         425215 non-null  float64
 17  ME

In [5]:
# Target variables for regression and classification
y_regression = df['MENTHLTH']
y_classification = df['MENT14D_binary']

In [6]:
# Feature matrices for regression and classification
X_A = df.drop(['MENTHLTH',  '_MENT14D', 'MENT14D_binary'], axis=1)
X_B = df.drop(['MENTHLTH',  '_MENT14D', 'MENT14D_binary', 'ADDEPEV3'], axis=1)

print(X_A.shape)
print(X_B.shape)

(425215, 42)
(425215, 41)


Model A — includes ADDEPEV3 (depression diagnosis) + all other predictors

Model B — excludes ADDEPEV3, predicts mental distress from sociodemographic and behavioral factors only

In [7]:
X_A_train, X_A_test, y_A_train, y_A_test = train_test_split(X_A, y_regression, test_size=0.2, random_state=42)
X_B_train, X_B_test, y_B_train, y_B_test = train_test_split(X_B, y_regression, test_size=0.2, random_state=42)

print(X_A_train.shape, X_A_test.shape)
print(X_B_train.shape, X_B_test.shape)

(340172, 42) (85043, 42)
(340172, 41) (85043, 41)


In [ ]:
# Linear Regression Model A / includes depression diagnosis

# Instantiate
lr_A = LinearRegression()

# Fit
lr_A.fit(X_A_train, y_A_train)

# Predict
y_A_pred = lr_A.predict(X_A_test)

In [ ]:
rmse_A = np.sqrt(mean_squared_error(y_A_test, y_A_pred))
mae_A = mean_absolute_error(y_A_test, y_A_pred)
r2_A = r2_score(y_A_test, y_A_pred)

print('Linear Regression Model A RMSE: ', rmse_A)
print('Linear Regression Model A MAE: ', mae_A)
print('Linear Regression Model A R2: ', r2_A)


Model A RMSE:  7.003800166963823
Model A MAE:  4.572315269528019
Model A R2:  0.28278964974144427


In [ ]:
# Linear Regression Model B / excludes depression diagnosis

# Instantiate
lr_B = LinearRegression()

# Fit
lr_B.fit(X_B_train, y_B_train)

# Predict
y_B_pred = lr_B.predict(X_B_test)

In [ ]:
rmse_B = np.sqrt(mean_squared_error(y_B_test, y_B_pred))
mae_B = mean_absolute_error(y_B_test, y_B_pred)
r2_B = r2_score(y_B_test, y_B_pred)

print('Linear Regression Model B RMSE: ', rmse_B)
print('Linear Regression Model B MAE: ', mae_B)
print('Linear Regression Model B R2: ', r2_B)


Model B RMSE:  7.384787214125774
Model B MAE:  4.957709720634669
Model B R2:  0.20263893026725865


* **RMSE**: Model A 7.00 vs Model B 7.38 — Model A is better
* **MAE**: Model A 4.57 vs Model B 4.96 — Model A is better
* **R²**: Model A 0.28 vs Model B 0.20 — Model A explains 8% more variance

Removing depression negatively impacts model performance; a depression diagnosis contributes meaningfully to predicting poor mental health days beyond what sociodemographic and behavioral factors alone can explain.

Linear Regression assumes a linear relationship between predictors and the outcome. Given the heavily skewed distribution of MENTHLTH seen in EDA, this assumption may not hold perfectly so will build a Random Forest model to compare.

In [ ]:
# Random Forest Model A / includes depression diagnosis

rf_A = RandomForestRegressor(n_estimators=100, random_state=42)
rf_A.fit(X_A_train, y_A_train)
y_A_pred_rf = rf_A.predict(X_A_test)

In [17]:
rmse_rf_A = np.sqrt(mean_squared_error(y_A_test, y_A_pred_rf))
mae_rf_A = mean_absolute_error(y_A_test, y_A_pred_rf)
r2_rf_A = r2_score(y_A_test, y_A_pred_rf)

print('Random Forest Model A RMSE: ', rmse_rf_A)
print('Random Forest Model A MAE: ', mae_rf_A)
print('Random Forest Model A R2: ', r2_rf_A)

Random Forest Model A RMSE:  7.128720765962655
Random Forest Model A MAE:  4.59319876086568
Random Forest Model A R2:  0.25697699031940147


In [18]:
# Random Forest Model B / excludes depression diagnosis

rf_B = RandomForestRegressor(n_estimators=100, random_state=42)
rf_B.fit(X_B_train, y_B_train)
y_B_pred_rf = rf_B.predict(X_B_test)

In [19]:
rmse_rf_B = np.sqrt(mean_squared_error(y_B_test, y_B_pred_rf))
mae_rf_B = mean_absolute_error(y_B_test, y_B_pred_rf)
r2_rf_B = r2_score(y_B_test, y_B_pred_rf)

print('Random Forest Model B RMSE: ', rmse_rf_B)
print('Random Forest Model B MAE: ', mae_rf_B)
print('Random Forest Model B R2: ', r2_rf_B)

Random Forest Model B RMSE:  7.584340128649748
Random Forest Model B MAE:  5.017495920997199
Random Forest Model B R2:  0.15896385984188066


| Model | RMSE | MAE | R² |
|---|---|---|---|
| Linear Regression A (with ADDEPEV3) | 7.00 | 4.57 | 0.283 |
| Linear Regression B (without ADDEPEV3) | 7.38 | 4.96 | 0.203 |
| Random Forest A (with ADDEPEV3) | 7.13 | 4.59 | 0.257 |
| Random Forest B (without ADDEPEV3) | 7.58 | 5.02 | 0.159 |

Neither model is performing particularly well on regression, which itself is a meaningful finding. It suggests that predicting the exact number of poor mental health days from survey data alone is genuinely difficult.

**Key Takeaways**
* A depression diagnosis is strongly related to a higher number of poor mental health days.
* The features selected for regression modeling do not strongly predict the number of poor mental health days.
    * The extreme skew in MENTHLTH makes regression inherently difficult
    * Self-reported survey data is noisy by nature
    * Many factors influence mental health that aren't captured in BRFSS, such as life events, trauma, relationships, etc
* Linear Regression marginally outperformed Random Forest on Model A (depression included) which suggests that the relationship between the predictors and MENTLTH are largely linear in nature.